### ISSUE TITLE
Be smart about autograd formulas saving either the input or output, depending on context

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#136733]
### 🚀 The feature, motivation and pitch

See https://github.com/NVIDIA/apex/pull/1715

The general idea is that for some operators, in principle the autograd formula can be written depending on *either* the input or the output.

For example, for the norm, usually `x_hat` is computed by saving the mean and rstd from the forwards, and then computing `(input - mean)  * rstd`. However, we can also compute it by recomputing it from the *output*, such as by doing `(output - bias) / weight`.

Which one is better is actually somewhat tricky to determine. For example, if the subsequent op requires the output of the norm  (like a matmul), then it would be better to save the output (i.e. `matmul(norm(x))`). On the other hand, if we already need to save the input of the norm, then saving the output would just require us to save more than necessary.

The general problem here is quite tricky, as it relies on "mathematical properties" (which autograd formulas can be rewritten?). But, I think if we choose a set of patterns we'd like to potentially rewrite, we could:

1. Do it as a pattern-matching pass on the joint graph.
2. Upon the pattern hitting, check if the input node or the output node already has other users. If so, we use the autograd formula that allows us to "reuse" the existing saved node.

**Comment by ngimel** [comment]
also see https://github.com/jax-ml/jax/pull/3232

**Comment by Chillee** [comment]
@ngimel I think it's a bit different, in that we want to make a choice about which autograd formula, while that PR is just about annotating a function with "invertible" to use a different autograd formula.

**Comment by Chillee** [comment]
I wrote a POC here: https://gist.github.com/Chillee/b6518e194b57d3bfbd2584fb5ea0df15 

<img width="606" alt="image" src="https://github.com/user-attachments/assets/89fef442-c895-42d2-9030-380ab3654ff3">

**Comment by ezyang** [comment]
Related? https://github.com/pytorch/pytorch/issues/23756

cc @vadimkantorov

**Comment by vadimkantorov** [comment]
- Yes, LayerNorm can save massive memory by not saving input for backward, saving just the rstd is sufficient (as discussed in https://github.com/pytorch/pytorch/issues/26288#issuecomment-1793209568) and even doing the computation inplace (same for BatchNorm and maybe others)

- In general, I think more of frequent ops should support both inplace/out modes and inverse formulas (https://github.com/pytorch/pytorch/issues/23756, https://github.com/pytorch/pytorch/issues/26288)

- As recomputation may have different/worse numerics, especially in low precision (and this is important since training is concerned), I personally always think that allowing nice user-controllable "plumbing" to choose which mode they want: maybe add publiclly callable inverse formulas, InverseSequential or some InversePreferringModule, so that the users can explicitly wrap and opt-in this memory-saving-doing-more-inplace-and-more-of-input-recomputation regime, or at least maybe there should be an explicit torch.compile-flag or context manager or some explicit `layer_norm(x).use_output_for_recomputing_input()` pseudo-functions for some explicit opt-in

- Also would be nice to have in docs code examples of frequent module blocks (e.g. Conv+BatchNorm+Relu+Dropout https://github.com/pytorch/pytorch/issues/92927) which are known and tested to be low-memory and doing inplace/recomputing, so that the user can at least copy-paste these blocks into their own code
